src/transport/client.rs

```diff
@@ -0,0 +1,37 @@
+use crate::{
+    common::Endpoint,
+    error::{Error, Result},
+    transport::{proto::RequestKind, Request, Response},
+};
+use tokio_sync::{mpsc, oneshot};
+
+pub type Channel = mpsc::Sender<(Request, oneshot::Sender<crate::Result<Response>>)>;
+
+#[derive(Debug, Clone)]
+pub struct Client {
+    inner: Channel,
+}
+
+impl Client {
+    pub fn new(inner: Channel) -> Self {
+        Self { inner }
+    }
+
+    pub async fn send(&mut self, target: Endpoint, req: RequestKind) -> Result<Response> {
+        let (tx, rx) = oneshot::channel();
+        let req = Request::new(target, req);
+
+        self.inner
+            .send((req, tx))
+            .await
+            .map_err(|_| Error::new_broken_pipe(None))?;
+
+        rx.await
+            .map_err(|_| Error::new_broken_pipe(None))?
+            .map_err(Into::into)
+    }
+
+    pub async fn broadcast(&mut self, req: RequestKind) -> Result<()> {
```

Comment:
An idea for how to impl this is to add a `endpoints: Vec<Endpoints>` parameter to this method. In the rapid impl, the endpoints on the `broadcaster` instance are a set of the membership set, which is updated in the MembershipService constructor:

https://github.com/lalithsuresh/rapid/blob/ec079d23640b71bf80968e26b8c00a97be1390c8/rapid/src/main/java/com/vrg/rapid/MembershipService.java#L146

And the broadcaster simply uses the regular client and makes the request to each.

Reply:
I think for broadcast what we want to do is actually enqueue broadcasts messages into a global queue, that then get processed via the cluser in the select!. This would then allow us to borrow the _current_ memebership view and broadcast out intead of having to replicate the membership view to each client.